benchopt/base.py

```diff
@@ -392,7 +392,11 @@ def set_dataset(self, dataset):
             parameters[key] = getattr(self, key)
         self.set_data(**data)
         for key in self._parameters:
-            if parameters[key] != getattr(self, key):
+            has_changed = (parameters[key] != getattr(self, key)
+                           if not hasattr(parameters[key], '__iter__')
+                           else any(parameters[key] != getattr(self, key)))
```

Comment:
Actually I think we can even simplify:
```suggestion
            # protect for cases where parameters[key] are numpy arrays.
            has_changed = any(parameters[key] != getattr(self, key))
```

Note that thinking back, this would not work for 2D ndarray.

Another way would be:
```python
has_changed = parameters[key] != getattr(self, key)
if not isinstance(has_changed, bool):
     has_changed = has_changed.any()
```

But not sure if it cover all potential cases.

Reply:
### for the fist suggestion
>``has_changed = any(parameters[key] != getattr(self, key))``

This would fail in the basic case

```python
any(1 != 2)  # TypeError: 'bool' object is not iterable 
```

### for the second suggestion

There is nothing that prevents packages from creating their own variable types (e.g numpy has its own bool type ``np.bool_``)

Therefore the second suggestion would fail also

```python
parameters = {
    'a': np.bool_(1),
    'b': np.bool_(0),,
}

isinstance(parameters['a'] != parameters['b'], bool)  # Output: False

# Applying the second solution would raise an exception 
# TypeError: 'numpy.bool_' object is not iterable
```

### Another suggestion

I believe that checking ``__iter__`` is less prone to bugs. I suggest a mixture between your solution and mine

```python
has_changed = parameters[key] != getattr(self, key)
if hasattr(has_changed, '__iter__'):
     has_changed = any(has_changed)
```

Reply:
Aha for the first one, I got owned by the use of `ipython` which imports symbols from `numpy`.
Indeed the basic `any` does not work with non iterable.

Let's go with the last suggestion then.